### PR TITLE
[mlir][python] Reland - Add sugared builder for transform.named_sequence

### PR DESCRIPTION
**Pull request by nicolasvasilache**
This reverts, addresses issues with #71597 and relands

### CODE REVIEW

File: mlir/test/python/dialects/transform.py [hunk 1]
```diff
@@ -10,13 +10,13 @@ def run(f):
         module = Module.create()
         with InsertionPoint(module.body):
             print("\nTEST:", f.__name__)
-            f()
+            f(module)
```
[the remark this answers is not in the document]
Comment (ftynse) [reply mlir/test/python/dialects/transform.py:13]: It's a choice between polluting the output of the test vs polluting the test code. I can live with any choice here.

**Review by ftynse — APPROVED** [review]
Note that merging via GH will just squash everything into one commit that will be seen as adding tests.

File: mlir/test/python/dialects/transform.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply mlir/test/python/dialects/transform.py:13]: ack, thanks!

**Comment by nicolasvasilache** [comment]
Pushed as:
2859a9cdfa633a38bbed75885b10203f6b9166fb -> revert
cb3880515fe9065ae3378f3d145a17bf75fa6740 -> revert
af3d85694427257ee27a71007af72ea11d54a93a -> fixed reland

**Pull request closed without merging**